pkg/rebuild/maven/maveninfer_test.go

```diff
@@ -257,3 +263,187 @@ func addPomArtifact(mavenRegistry *mockMavenRegistry, pom *PomXML) {
 	xmlBytes, _ := xml.MarshalIndent(pom, "", "  ")
 	mavenRegistry.artifactCoordinates[key] = xmlBytes
 }
+
+func TestMavenInfer(t *testing.T) {
+	testCases := []struct {
+		name              string
+		target            rebuild.Target
+		repo              string
+		zipEntries        map[string][]*archive.ZipEntry
+		expectedHeuristic string
+		wantErr           bool
+	}{
+		{
+			name: "infer using git log heuristic",
+			target: rebuild.Target{
+				Ecosystem: "Maven",
+				Package:   "foo:bar",
+				Version:   "1.0.0",
+			},
+			repo: `
+            commits:
+              - id: initial-commit
+                files:
+                  pom.xml: |
+                    <project>
+                        <modelVersion>4.0.0</modelVersion>
+                        <groupId>foo</groupId>
+                        <artifactId>bar</artifactId>
+                        <version>1.0.0</version>
+                    </project>`,
+			zipEntries: map[string][]*archive.ZipEntry{
+				maven.TypeSources: {{
+					FileHeader: &zip.FileHeader{Name: "Foo.java"},
+					Body:       []byte("class Foo {}"),
+				}},
+				maven.TypeJar: {
+					{
+						FileHeader: &zip.FileHeader{Name: "META-INF/maven/foo/bar/pom.xml"},
+						Body:       []byte(`<project><groupId>foo</groupId><artifactId>bar</artifactId><version>1.0.0</version></project>`),
+					},
+					{
+						FileHeader: &zip.FileHeader{Name: "META-INF/MANIFEST.MF"},
+						Body:       []byte("Manifest-Version: 1.0\nBuild-Jdk: 11.0.1\n"),
+					},
+				},
+			},
+			expectedHeuristic: "using git log heuristic (pkg and version match)",
+			wantErr:           false,
+		},
+		{
+			name: "infer using source jar heuristic",
+			target: rebuild.Target{
+				Ecosystem: "Maven",
+				Package:   "foo:bar",
+				Version:   "1.0.0",
+			},
+			repo: `
+            commits:
+              - id: initial-commit
+                files:
+                  pom.xml: |
+                    <project>
+                        <modelVersion>4.0.0</modelVersion>
+                        <groupId>foo</groupId>
+                        <artifactId>bar</artifactId>
+                        <version>0.0.0-dev</version>
+                    </project>
+                  src/main/java/Foo.java: |
+                    a`,
+			zipEntries: map[string][]*archive.ZipEntry{
+				maven.TypeSources: {{
+					FileHeader: &zip.FileHeader{Name: "src/main/java/Foo.java"},
+					Body:       []byte("a"),
+				}},
+				maven.TypeJar: {
+					{
+						FileHeader: &zip.FileHeader{Name: "META-INF/maven/foo/bar/pom.xml"},
+						Body:       []byte("<project><groupId>foo</groupId><artifactId>bar</artifactId><version>1.0.0</version></project>"),
+					},
+					{
+						FileHeader: &zip.FileHeader{Name: "META-INF/MANIFEST.MF"},
+						Body:       []byte("Manifest-Version: 1.0\nBuild-Jdk: 11.0.1\n"),
+					},
+				},
+			},
+			expectedHeuristic: "using source jar heuristic with mismatched version",
```

Comment:
This isn't ideal imo since we really want to be checking the program outputs, not its incidental side-effects. that said, we can revisit later